Update footer save warning to use the Stimulus CloneController (w-messages)

lb- opened the pull request:
* Fixes #10091
* This work is a partial enabler for #8453 (making it easier to use this behaviour in other forms) and gets us very close to being able to pull out page-editor.js (after this PR, there will only be about 4 lines of code in that file), see #2936 for context.

## Details

* The initial commit renames MessagesController to CloneController and keeps the identifier `w-messages` but also adds `w-clone`. This is because this controller is better named by the behaviour it sets up, not where it is used (making message content show), further Stimulus plans involve using this controller for its generic behaviour.
* The second commit is a refactor of the work stated in #10327 (this PR replaces that PR), the initial starting point was there but needed a fair bit of work to get to a functional state.
- Controller changes include
  - Changes default show-delay to zero (immediate) and update the w-messages (page messages) to have the 100ms delay
  - Add support for 'hide' classes to be added/removed based on clone usage
  - Add support for calling the clear method from actions, and resetting classes if used after an animation duration or max delay
  - Pull out logic to get template content to class method
  - General clean up of the original messages only implementation to be better documented, typed and tested, bringing in the learnings from other Controller work
- Refine the `debounce` util to be more useful, updating Types to be better, now returns a Promise API as per the planned implementation for #10356 and also allows this to be used without a delay. This means the one util can be used to conditionally immediately run the callback. Super useful when we want the same code to run maybe delayed as we are doing in a few of the Stimulus controllers.
- Finally, this adopts the w-messages (CloneController) in unsaved changes warning, replacing the need for the global `window.updateFooterSaveWarning` and now uses Event dispatching to perform the same behaviour.
- I have added `aria-live="polite"` so that screen readers may advise users of these messages.


## Notes on scope

* Any core functional changes to the `enableDirtyFormCheck` function, this is what runs to compare the initial form state to the potentially changed state and sets up comment changes listeners. There are known issues with this code but at the moment this PR is not trying to refine/resolve any of them, behaviour should just be the same as current state.
* Backwards compatible global `updateFooterSaveWarning`, can add this if needed, but this felt like an internal function and would not be called by custom code.
* I added `{% block message_templates %}`, I thought this may be a nice way to add more unsaved pages messages through customisations, this is intentionally not documented.


## Checklist

-   [X] Do the tests still pass?
-   [X] Does the code comply with the style guide?
-   [X] For Python changes: Have you added tests to cover the new/fixed behaviour? (N/A)
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments? Chrome 118, Firefox 115, Safari 16.1
-   [X] For new features: Has the documentation been updated accordingly? (added release considerations


## Testing

Check these items in both a new Page creation form and an existing Page edit form. Note that the page create form acts a bit differently with how it calculates whether the page has edits. See also #7188 .

1. Check that when the page form loads, after ~3 seconds, no rogue 'unsaved edits' appears except for know issue #1234
2. Now make an edit to any basic field, observe that the unsaved field edits message appears in the footer
3. Now add a comment to any field, observe the message should change to field & comments in the footer
4. Remove the comment, observe the message updates
5. You can also remove the change made in step 2 (basic field change), set it back to the previous, the message should remove after (~2s)
6. Refresh, do not save anything, now just add a comment somewhere. Observe the message should appear.
7. Now remove that comment, observe after a delay (~2s) the message should be removed
8. **additional test** Add a comment, observe the message shows, now remove the comment and quickly add one, wait ~2 seconds and confirm the message is still shown. (this checks the setTimeout tracking).

lb- commented on client/src/controllers/CloneController.ts:
```diff
@@ -0,0 +1,148 @@
+    const content = template.content.firstElementChild?.cloneNode(true);
+    return content instanceof HTMLElement ? content : null;
+  }
+}
```
Note: Unfortunately Git diffs do not do renamed files well. However the MessagesController has been renamed and while a lot has changed the core behaviour is still there. When reviewing the commits locally it should be a bit clearer what changed.

lb- commented on wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html:
```diff
@@ -1,11 +1,32 @@
+    data-w-messages-hide-class="footer__container--hidden"
+>
+    <div class="footer__save-warning footer__emphasise-span-tags" data-w-messages-target="container"></div>
+    {% trans 'Save the page before leaving' as generic_save_page_message %}
```
Note: If we want, I can pull this 'shared' label up into the data-w-messages-text-param and remove from each of the `p` tags below. Not sure if that will make this HTML readability worse though.

laymonage replied:
I don't think it's necessary, though perhaps it should be inside the `message_templates` block below, so if we later extract this into a shared template, we'll literally just have the block's contents in this file, e.g.

```html+django
{% extends "wagtailadmin/shared/unsaved_changes_warning.html" %}
{% block message_templates %}
    {% trans 'Save the page before leaving' as generic_save_page_message %}
    {# The rest of the templates here ... #}
{% endblock %}
```

lb- commented on wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html:
```diff
@@ -1,11 +1,32 @@
-    </div>
+<li
+    class="footer__container footer__container--hidden"
+    role="status"
```
```suggestion
    aria-live="polite"
```

lb- commented:
Just rebased this and removed some code that was not really required for this feature adoption.

laymonage commented on client/src/entrypoints/admin/page-editor.js:
```diff
@@ -24,43 +24,13 @@ function initKeyboardShortcuts() {
+ *
+ * @deprecated RemovedInWagtail60
+ */
+window.updateFooterSaveWarning = () => {};
```
> Backwards compatible global `updateFooterSaveWarning`, can add this if needed, but this felt like an internal function and would not be called by custom code.

I don't think it's necessary, this stub function is more than enough 👍

lb- replied:
Great, I will remove this.

laymonage reviewed (approved):
Tested all the listed scenarios and happy with the overall changes. Thanks for keeping the PR scope focused so it's easy to review!

Just have a minor comment on moving the translatable message string to inside the newly-introduced block. Otherwise it's all good 👍

lb- commented:
Thanks @laymonage  I will merge this in.

I have moved that translation into inside template block, good call.

I have removed the noop stub of the removed function and updated the release considerations wording accordingly.

Added changelog and will merge once the CI passes.

Thanks @Hussain0520 for your original start here, really appreciate it.
